[RFC] Refactor API dispatch, adding lua templating engine

ZyX-I opened the pull request:
Current lua code is not readable, using templating engine instead is better. Though I needed a number of hacks to make the result look fine (i.e. have nice indentation and empty lines), it is still better then what there was.

Also note import from test helpers, maybe I should move the code elsewhere instead and import it *in* test helpers.

ZyX-I committed:
Add lust template engine

ZyX-I committed:
oldtests: Remove all swap files, which includes things like .swn

ZyX-I committed:
dispatch: Replace unreadable lua code with lust template

Also gets rids of useless Map(String, MsgpackRpcRequestHandler): khash interface 
is already fine as-is for just “put” and “get” and Map(…) does useless job of 
double allocating memory where no allocations are needed at all, up to first 
allocations needed to start filling the data.

Possible TODO: use perfect hash instead.

ZyX-I commented:
Currently there were two intentional changes to dispatch code structure:

1. All parameter declarations were moved to near where parameters are checked.
2. Special handle types type check was merged with integer type check. Do not know why they were separate.

ZyX-I commented:
Though no, second looks like a refactoring artifact from times when there were `Buffer buffer`, etc in the union.

justinmk commented:
Why we are we spending time on this? The old code wasn't that hard to read, not to mention it probably won't change very often. 

The new template thing is a new syntax, and now I have to find out what `template:register()`  does ...

ZyX-I commented:
@justinmk Because we are generating many different things, meaning much of unreadable code. I wanted to have something like this back when I was writing lua bindings. And code I changed is one of the least easy to modify things in Neovim (especially if you want to keep indentation fine), it is only “not that hard to read” because there are comments which are not needed with proper templates. And in any case my template breaks 80 column limit (in template code, not in the generated code) in one place only, not in fifteen places.

ZyX-I committed:
dispatch: Fix number in “wrong type for argument N” error message

ZyX-I committed:
dispatch: Also convert integers to booleans

This conversion was in previous variant, removal caused some test failures.

ZyX-I commented:
@justinmk What I mean is that for new generators it would be good to have a templating engine. But it would not be used without making old generators use it in advance: many code is written in a fashion “check how previous code solved similar task and copy what is possible from there”.

ZyX-I committed:
dispatch: Use shorter separator syntax

This way 80 character limit is not broken.

justinmk commented:
> But it would not be used without making old generators use it in advance: many code is written in a fashion “check how previous code solved similar task and copy what is possible from there”.

That's true, but it also makes it harder to reverse the decision. And we can see that there is risk of new bugs.

> there are comments which are not needed with proper templates

Personally I still would find the comments valuable.

ZyX-I committed:
dispatch: Remove parenthesis, some compilers can’t swallow them

ZyX-I commented:
@justinmk I would not ever want to stop using templates there. And switching template systems is much easier when it is *switching* and not refactoring raw lua code into template.

All comments I saw there are “obvious” ones. Some are even obvious with the current code, others are not needed with proper variable and subtemplates naming.

bfredl commented:
I wonder how it will work in a case like `ui_events.lua`. Here the "primitive" approach allows us to group related code together (like vtable declaration and implementation) even if they are put in different files.

ZyX-I commented:
@bfredl One can just use subtemplates to reuse same code pieces. Not that there is much code to share between files, primary signatures.

ZyX-I committed:
dispatch: Remove unneeded includes

ZyX-I committed:
dispatch: Make dispatch_wrappers a *.c and not *.h file

Allows to specify headers with C code, removes explicit API headers list to 
maintain, makes code more consistent with lua bindings.

ZyX-I committed:
dispatch: Add function attributes to handlers

ZyX-I committed:
dispatch: Make lua bindings use template engine

ZyX-I committed:
dispatch: Make output of new and previous version almost the same

Checked with

    diff -u <(luajit src/nvim/generators/gen_api_dispatch.lua src/nvim /dev/null /dev/null /dev/null /dev/stdout src/nvim/api/*.h build/include/api/*.h | grep -v '^$' | perl -p -e 's/\brv\b/ret/g') <(cat build/src/nvim/auto/msgpack_lua_c_bindings.generated.c|grep -v '^$')

Some changes are to be reverted to make output more consistent with the rpc 
dispatch output.

ZyX-I committed:
dispatch: Make dispatch and lua bindings more consistent

Now one may use

    diff -u \
      <(luajit src/nvim/generators/gen_api_dispatch.lua \
               src/nvim \
               /dev/null /dev/null /dev/null /dev/stdout \
               src/nvim/api/*.h build/include/api/*.h \
        | grep -vFx '' \
        | grep -vFx '  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT' \
        | perl -p -e 's/\brv\b/ret/g;s/\barg_(\d+)\b/"arg".($1+1)/ge' \
        | perl -e '($_ = join("", <>));s/\n\s*(?="Expected)/ /g;print') \
      <(cat build/src/nvim/auto/msgpack_lua_c_bindings.generated.c \
        | grep -v '^$')

to check they are the same, given …/msgpack_lua_c_bindings.generated.c
contains output from previous version of the generator.

ZyX-I committed:
dispatch: Remove unused functions

ZyX-I committed:
generators: Make gen_char_blob.lua use dump_bin_array.lua

Also refactors the latter.

ZyX-I committed:
dispatch: Rename output to dispatch_output

ZyX-I committed:
eval: Move functions perfect hash generating code to gperf.lua

To be used later for dispatch.

ZyX-I committed:
dispatch: Use gperf in place of khash

ZyX-I committed:
gperf: Do not expect NUL-terminated strings

bfredl commented on src/nvim/map.c:
```diff
@@ -123,29 +123,8 @@
     kh_clear(T##_##U##_map, map->table); \
   }
 
-static inline khint_t String_hash(String s)
```
Is there really a point to delete these? It seems not very unlikely that we might want a mutable string hash in the future.

ZyX-I replied:
@bfredl From `map.c`? There is a point to delete `map.[ch]` entirely, ineffective third hash interface which just indirects second hash interface which uses second hash implementation is not a thing we really need. I was expressing the reasons previously.

Initially I did not delete String_hash, I moved it to `dispatch.c`. It was then deleted when I decided that dispatch is a perfect case for gperf. If you think it can be needed I can move the function to `khash.h`, but I do not see any reasons to keep anything in `map.c` I can remove instead.

bfredl replied:
I don't mind removing the indirection, even if we have helper functions/macros they can just operate on the khash types directly. But anyway I guess this could be in `khash.c` as it is khash that receives a hash function.

ZyX-I replied:
I am wondering whether `String_hash` stopping at NUL byte, so that `abc\0def` and `abc\0ghi` have the same hash is intentional?

ZyX-I commented:
Got test failures like

```
[ RUN      ] api nvim_get_mode during normal-mode `g` returns blocking=true: 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

Would say that this looks unrelated, but I got them everywhere, so it should be something else. Will first rebase to check whether failures go away.

ZyX-I commented:
Though it is reproducible locally without rebase.

ZyX-I commented:
Gperf generated code is even more inadequate, debugging symbols point to incorrect location because of its `#line` directives.

bfredl commented:
> api nvim_get_mode during normal-mode `g` returns blocking=true: 

That sounds like trouble with the `async` flag, but at a first glance it looks you handle it correctly. To be sure you could inspect the gperf output generated for `nvim_get_mode`

ZyX-I commented:
I found it, `async` flag got missing for some reason.

Got it, I have accidentally replaced `fn.async` with `async`. Damn that “missing var is nil” semantics.

ZyX-I committed:
dispatch: Check async flag correctly

ZyX-I committed:
khash: Add binary string hashes

Not actually used anywhere.

ZyX-I commented:
Should be more or less ready now, last commit adds code which does not affect anything, previous commit has passing CI.

ZyX-I commented:
Note that in the last commit `&& s.data[i]` was removed, making `kh_bin_str_hash_func` always traverse the whole binary string.

justinmk commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -4,43 +4,65 @@ mpack = require('mpack')
-for i = 6, #arg do
-  local full_path = arg[i]
+local dedent = global_test_helpers.dedent
+local shallowcopy = global_test_helpers.shallowcopy
```
should these be moved out of `test/helpers.lua` ?

justinmk replied:
Exploring this in https://github.com/neovim/neovim/pull/9301.

ZyX-I commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -4,43 +4,65 @@ mpack = require('mpack')
-for i = 6, #arg do
-  local full_path = arg[i]
+local dedent = global_test_helpers.dedent
+local shallowcopy = global_test_helpers.shallowcopy
```
@justinmk It is what I was asking in some comment: should it be moved and where if yes.

justinmk replied:
@ZyX-I `src/nvim/lib/util.lua` makes the most sense to me. Can break it out in to `src/nvim/lib/{foo,bar}.lua` later if it gets big.

ZyX-I commented:
QB timeout looks unrelated:

```
10:49:36,709 INFO  - # test/functional/ex_cmds/cd_spec.lua @ 282: getcwd() returns empty string if working directory does not exist
10:49:36,713 INFO  - ok 815 - getcwd() returns empty string if working directory does not exist
10:49:36,725 WARN  - 2017/08/12 10:49:36 ERROR 75866/open_log_file:174: Logging to stderr, failed to open $NVIM_LOG_FILE: .nvimlog
10:49:36,725 WARN  - 2017/08/12 10:49:36 WARN 75866/call_set_error:815: RPC: ch 1 was closed by the client
10:49:36,725 INFO  - # test/functional/ex_cmds/ctrl_c_spec.lua @ 11: CTRL-C (mapped) interrupts :global
10:55:06,252 INFO  - Terminating launched command gracefully...
10:55:06,276 WARN  - Vim: Caught deadly signal 'SIGTERM'
10:55:06,276 WARN  -
10:55:06,276 WARN  - Vim: Finished.
10:55:06,277 INFO  - not ok 816 - CTRL-C (mapped) interrupts :global
10:55:06,277 INFO  - # test/functional/ex_cmds/ctrl_c_spec.lua @ 11
10:55:06,277 INFO  - # Failure message: ./test/functional/helpers.lua:191: attempt to perform arithmetic on local 'written' (a nil value)
10:55:06,277 INFO  - # stack traceback:
10:55:06,277 INFO  - #     ./test/functional/helpers.lua:191: in function 'nvim_feed'
10:55:06,277 INFO  - #     ./test/functional/helpers.lua:197: in function 'feed'
10:55:06,277 INFO  - #     test/functional/ex_cmds/ctrl_c_spec.lua:43: in function 'test_ctrl_c'
10:55:06,277 INFO  - #     test/functional/ex_cmds/ctrl_c_spec.lua:56: in function <test/functional/ex_cmds/ctrl_c_spec.lua:11>
10:55:06,277 INFO  - #
10:55:06,282 WARN  - Terminated
10:55:06,282 WARN  - Terminated
```

Travis hanged at test49 again:

```
shell returned 1
</build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/136" </neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/136" 0L, 0C"~/build/neovim/neovim/src/nvim/testdir/test49.vim" <eovim/neovim/src/nvim/testdir/test49.vim" [readonly] 9009L, 206897C"test49.in" "test49.in" 31L, 668C</build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/126" <eovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/126" 6L, 216C"test49.in" "test49.in" 31L, 668C"~/build/neovim/neovim/src/nvim/testdir/test49.vim" <eovim/neovim/src/nvim/testdir/test49.vim" [readonly] 9009L, 206897C
:!echo 'source /home/travis/build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/1' >/home/travis/build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137
/^\s*if\s\+ExtraVim(.*)/+/^\s*if\s\+ExtraVim(.*)/+1</build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" <ovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" 5L, 62C appended</build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" <eovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" 6L, 145C
5 substitutions on 5 lines</build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" <ovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/137" 6L, 140C written"~/build/neovim/neovim/src/nvim/testdir/test49.vim" <eovim/neovim/src/nvim/testdir/test49.vim" [readonly] 9009L, 206897C
:!>/home/travis/build/neovim/neovim/src/nvim/testdir/Xtest-tmpdir/nvimEaf4AV/138
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

ZyX-I committed:
tests,dispatch: Move some functions to src/nvim/lib/helpers.lua
